modules/core/core-internal/src/main/java/com/enonic/xp/core/internal/FileNames.java

```diff
@@ -0,0 +1,100 @@
+package com.enonic.xp.core.internal;
+
+import java.nio.charset.StandardCharsets;
+import java.text.Normalizer;
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class FileNames
+{
+    // From Windows File Naming Conventions: reserved characters <>:"/\|?*
+    // Sorted for binary search
+    private static final int[] RESERVED_CHARACTERS = "<>:\"/\\|?*".chars().sorted().toArray();
+
+    // From Windows File Naming Conventions, extended with COM0 and LPT0 (they are not acceptable either)
+    public static final Set<String> RESERVED_NAME =
+        Set.of( "con", "prn", "aux", "nul", "com0", "com1", "com2", "com3", "com4", "com5", "com6", "com7", "com8", "com9", "lpt0", "lpt1",
+                "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9" );
+
+    public static final Set<String> RESERVED_PREFIX =
+        RESERVED_NAME.stream().map( reserved -> reserved + "." ).collect( Collectors.toUnmodifiableSet() );
+
+    // All types of invisible characters, including surrogate to disallow non-BMP characters
+    // Unassigned characters are can't be printed and also considered invisible
+    // Sorted for binary search
+    public static final int[] INVISIBLE_CHARACTER_TYPES =
+        IntStream.of( Character.UNASSIGNED, Character.SPACE_SEPARATOR, Character.LINE_SEPARATOR, Character.PARAGRAPH_SEPARATOR,
+                      Character.CONTROL, Character.FORMAT, Character.PRIVATE_USE, Character.SURROGATE ).sorted().toArray();
+
+    private FileNames()
+    {
+    }
+
+    public static boolean isSafeFileName( final String fileName )
```

Comment:
Shouldn't it return `false` for null?

Reply:
null is not unsafe name. there is no name. I change code to throw NPE more explicitly